src/System.Threading/tests/SemaphoreSlimTests.cs

```diff
@@ -0,0 +1,751 @@
+                }
+                if (semaphore.AvailableWaitHandle.WaitOne(0) != state)
+                {
+                    string methodFailed = "RunSemaphoreSlimTest7_AvailableWaitHandle(" + initial + "," + maximum + "," + action + "): FAILED.  ";
+                    Assert.True(false, string.Format(methodFailed + "AvailableWaitHandle failed, expected " + state + " actual " + !state));
+                }
+            }
+            catch (Exception ex)
+            {
+                string methodFailed = "RunSemaphoreSlimTest7_AvailableWaitHandle(" + initial + "," + maximum + "," + action + "): FAILED.  ";
+                Assert.True(false, string.Format(methodFailed + "AvailableWaitHandle failed, the code threw exception " + ex));
+            }
+        }
+
+        /// <summary>
+        /// Test SemaphoreSlim Wait and Release methods concurrently
+        /// </summary>
+        /// <param name="initial">The initial semaphore count</param>
+        /// <param name="maximum">The maximum semaphore count</param>
+        /// <param name="waitThreads">Number of the threads that call Wait method</param>
+        /// <param name="releaseThreads">Number of the threads that call Release method</param>
+        /// <param name="succeededWait">Number of succeeded wait threads</param>
+        /// <param name="failedWait">Number of failed wait threads</param>
+        /// <param name="finalCount">The final semaphore count</param>
+        /// <returns>True if the test succeeded, false otherwise</returns>
+        private static void RunSemaphoreSlimTest8_ConcWaitAndRelease(int initial, int maximum,
+            int waitThreads, int releaseThreads, int succeededWait, int failedWait, int finalCount, int timeout)
+        {
+            string methodFailed =
+                "RunSemaphoreSlimTest8_ConcWaitAndRelease("
+                + initial + "," + maximum + "," + waitThreads + ", " + releaseThreads + ", "
+                + succeededWait + ", " + failedWait + ", " + finalCount + ", " + timeout + "): FAILED.  ";
+            try
+            {
+                SemaphoreSlim semaphore = new SemaphoreSlim(initial, maximum);
+                Task[] threads = new Task[waitThreads + releaseThreads];
+                int succeeded = 0;
+                int failed = 0;
+                ManualResetEvent mre = new ManualResetEvent(false);
+                // launch threads
+                for (int i = 0; i < threads.Length; i++)
+                {
+                    if (i < waitThreads)
+                    {
+                        // We are creating the Task using TaskCreationOptions.LongRunning because...
+                        // there is no guarantee that the Task will be created on another thread.
+                        // There is also no guarantee that using this TaskCreationOption will force
+                        // it to be run on another thread.
+                        threads[i] = new Task(delegate ()
```

Comment:
Should use Task.Run, same for the one below (and then of course you don't need the subsequent .Start)